### PR TITLE
Clean up VmError

### PR DESCRIPTION
**Pull request by axic**
Pulled from #174.

### CODE REVIEW

**Comment by jwasinger** [comment]
I got this running.  45 failing tests.  I wonder if the VM is actually broken in those.

File: lib/runCode.js [hunk 1]
```diff
@@ -193,8 +193,7 @@ module.exports = function (opts, cb) {
         // run the opcode
         var result = opFn.apply(null, args)
       } catch (e) {
-        runState.vmError = e.error
-        cb()
+        cb(new VmError(e.error))
```
Comment (axic) [inline lib/runCode.js]: This actually could be `cb(e)`

**Comment by holgerd77** [comment]
This test run falls in the period where we temporarily had the Promise-version for async/await in ``ethereumjs-testing`` with the memory overflow problems, might have influenced the results. Have restarted the build, we'll see what comes out of this.

**Comment by holgerd77** [comment]
Ah no, still failing. :-/

**Comment by axic** [comment]
Looking at the below more carefully:
```
   function parseVmResults (err) {
     err = runState.vmError || err
```

Shows that `err` must be an error text ("constant") and not an exception object.

**Comment by holgerd77** [comment]
Can't investigate further - no time - but here are some results when running one failing test with ``node tests/tester -s --test='returndatacopy_following_revert' --jsontrace`` on both branches:

master
```
{"pc":41,"op":96,"gas":"0x9ffffab10","gasCost":"0x3","stack":["0x20"],"depth":0,"opName":"PUSH1"}
{"pc":43,"op":96,"gas":"0x9ffffab0d","gasCost":"0x3","stack":["0x20","0x0"],"depth":0,"opName":"PUSH1"}
{"pc":45,"op":62,"gas":"0x9ffffab0a","gasCost":"0x3","stack":["0x20","0x0","0x0"],"depth":0,"opName":"RETURNDATACOPY"}
{"pc":46,"op":96,"gas":"0x9ffffab04","gasCost":"0x3","stack":[],"depth":0,"opName":"PUSH1"}
{"pc":48,"op":81,"gas":"0x9ffffab01","gasCost":"0x3","stack":["0x0"],"depth":0,"opName":"MLOAD"}
{"pc":49,"op":96,"gas":"0x9ffffaafb","gasCost":"0x3","stack":["0x111122223333444455556666777788889999aaaabbbbccccddddeeeeffff"],"depth":0,"opName":"PUSH1"}
{"pc":51,"op":85,"gas":"0x9ffffaaf8","gasCost":"0x0","stack":["0x111122223333444455556666777788889999aaaabbbbccccddddeeeeffff","0x0"],"depth":0,"opName":"SSTORE"}
ok 1 the state roots should match

1..1
# tests 1
# pass  1
```

error-cb
```
{"pc":41,"op":96,"gas":"0xffffab22","gasCost":"0x3","stack":["0x20"],"depth":0,"opName":"PUSH1"}
{"pc":43,"op":96,"gas":"0xffffab1f","gasCost":"0x3","stack":["0x20","0x0"],"depth":0,"opName":"PUSH1"}
{"pc":45,"op":62,"gas":"0xffffab1c","gasCost":"0x3","stack":["0x20","0x0","0x0"],"depth":0,"opName":"RETURNDATACOPY"}
not ok 1 the state roots should match
  ---
    operator: equal
    expected: |-
      '7021b60fa4ce0fd56ce1b85e9c529345300b005574afb1f51a8363471cefdcd3'
    actual: |-
      'dbcee120f31bca7ddc3ec1d52a1b861c02c63d469e954096ff264858b58df101'
    at: replenish (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:946:17)
  ...

1..1
# tests 1
# pass  0
# fail  1
```

**Comment by Silur** [comment]
returnDataCopy revert only needed a deepcheck for `err`, currently failing
- RevertOpcodeCreate
- RevertOpcodeInCreateReturns
- RevertOpcodeInInit
fix in progress

File: lib/runCode.js [hunk 2]
```diff
@@ -214,7 +213,7 @@ module.exports = function (opts, cb) {
   }
 
   function parseVmResults (err) {
-    err = runState.vmError || err
+    err = runState.vmError || (err && (err.error || err))
```
Comment (axic) [inline lib/runCode.js]: Would be nice if the instances not returning an error object would be fixed instead. Do you happen to know which ones?

Comment (jwasinger) [reply lib/runCode.js]: Okay.  I was able to remove this.  now we are back to the failures that @Silur mentioned.

**Comment by jwasinger** [comment]
Hey, @Silur.  My apologies, I forgot you had claimed this issue.  Were you able to make any progress on the four failing tests?  Or can I try and get this finished?

**Comment by Silur** [comment]
oh great you found the part where it didn't return with an err!
I put the hackcheck for that, lemme see what breaks now

File: lib/runCode.js [hunk 3]
```diff
@@ -193,8 +193,7 @@ module.exports = function (opts, cb) {
         // run the opcode
         var result = opFn.apply(null, args)
       } catch (e) {
-        runState.vmError = e.error
```
Comment (Silur) [inline lib/runCode.js]: this is where `RevertOpcodeCreate` breaks, using `new VmError(e)` or `new VmError(e.error)` doesn't work, maybe leave the runState error here for e is not in any of the categorized exceptions?

File: lib/runCode.js [hunk 4]
```diff
@@ -193,6 +193,7 @@ module.exports = function (opts, cb) {
         // run the opcode
         var result = opFn.apply(null, args)
       } catch (e) {
+				runState.vmError = e
```
Comment (Silur) [inline lib/runCode.js]: any error occurred here does not belong to any of the known exceptions, so excluding them from the runState will break stateroots

File: lib/runCode.js (hunk 3 above)
Comment (jwasinger) [reply lib/runCode.js]: I'm not even sure if `vmError` is used anymore after these changes.  Maybe we can remove it.

File: lib/runCode.js (hunk 4 above)
Comment (jwasinger) [reply lib/runCode.js]: I see @axic has incorporated a new type of error `INTERNAL_ERROR` (https://github.com/ethereumjs/ethereumjs-vm/blob/35e8ddcc71ae47954d5a2599d784b3e63507df6f/lib/constants.js#L9) in https://github.com/ethereumjs/ethereumjs-vm/pull/222 .

Maybe we can pull those changes to this PR, and add a new field that allows us to specify a custom error message for `INTERNAL_ERROR`s .

File: lib/runCode.js (hunk 3 above)
Comment (axic) [reply lib/runCode.js]: `vmError` is used in the `result` output object by users of the VM (such as Remix)

Comment (jwasinger) [reply lib/runCode.js]: Ah okay.  Best to leave it in then :D

**Comment by axic** [comment]
@jwasinger didn't know it will dismiss the review :(

**Comment by axic** [comment]
@Silur @jwasinger thanks for fixing this!

File: lib/runCode.js (hunk 3 above)
Comment (axic) [reply lib/runCode.js]: Sorry, double checked remix and it uses `exceptionError`. Though there might be some other user which uses `vmError`, so I should add that back it properly

**Comment by axic** [comment]
@jwasinger @Silur please review the last commit

**Pull request merged**